File: apps/androidtv/feature/browse/src/main/kotlin/dev/spidola/tv/feature/browse/BrowseUiState.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package dev.spidola.tv.feature.browse

import kotlinx.collections.immutable.ImmutableList

/** The browse screen's state. A closed set, matched exhaustively by the composable. */
sealed interface BrowseUiState {
    data object Loading : BrowseUiState

    data object Empty : BrowseUiState

    data class Ready(val channels: ImmutableList<ChannelItem>) : BrowseUiState

    data class Error(val message: String) : BrowseUiState
}

/**
 * A single channel row's display data. [key] is the stable per-source identity (not the rowid),
 * so focus restoration survives a catalog refresh (Compose focusRestorer needs stable keys).
 */
data class ChannelItem(
    val key: Long,
    val name: String,
    val group: String?,
)

File: apps/tvos/Packages/FeatureBrowse/Sources/FeatureBrowse/BrowseUiState.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

/// The browse screen's state. A closed set, matched exhaustively by the view.
public enum BrowseUiState: Sendable {
  case loading
  case empty
  case ready([ChannelItem])
  case error(String)
}

/// A single channel row's display data. `id` is the stable per-source identity (not the rowid),
/// so it doubles as the SwiftUI list identity and the focus key across a catalog refresh.
public struct ChannelItem: Identifiable, Hashable, Sendable {
  public let id: Int64
  public let name: String
  public let group: String?

  public init(id: Int64, name: String, group: String?) {
    self.id = id
    self.name = name
    self.group = group
  }
}
